docs/dynamic-programming/coin-change-problem.mdx

```diff
@@ -0,0 +1,229 @@
+---
+title: Coin Change Problem
+slug: /dynamic-programming/coin-change-problem
+---
+
+## Introduction
+
+Given an integer `amount`, which represents the total amount of money, an integer array `coins` representing different coin values, and each denomination has an infinite supply of coins. Find what will be the minimum number of coins required to make up the given amount?
+
+## Example
+
+coins[]={1,5,6,9}
+
+n=4
+
+amount=10
+
+Final `dp` array after following the algorithmic steps:
+
+![image](https://user-images.githubusercontent.com/64724041/117053094-2bd2b300-ad36-11eb-87a6-ef2f6397440e.png)
+
+`Answer = dp[4][10] = 2`
+
+## Explaination
+
+In this problem, we are using the **Bottom-up approach** of dynamic programming. In the dynamic programming approach, we divide the problem into subproblems. We divide this problem into subproblems by dividing the amount from 0 to the total amount and representing it as the column-indices of the 2-d array. We will solve these subproblems and store them in cells of the 2-d array. The row-indices of this 2-d array represent indices of coins array.
+
+## Algorithm
+
+- Make a 2-d array, say `dp[n+1][amount+1]`.
+- Assign all the columns of the `0th` row of the 2d-array to a maximum value: `dp[0][i]= INT_MAX`.
+- Assign all the rows of the `0th` column of the 2-d array as 0, that is, `dp[i][0]=0`.
+- For the `1st` row, if the column indices value is divisible by the first coin, write the quotient in that cell, else write the max value.
+- For row>1, iterate till column index value is less than the ith coin value (i.e for all column indices, j,such that, j<coins[i-1] )and assign `dp[i][j]` to the previous upper row value, that is, `dp[i][j]=dp[i-1][j]`.
+- Now we have two options for all column indices, j, such that, j>=coins[i-1], whether to include ith coin or not.
+- Case1- If we don't consider the ith coin, then the cell value is the previous row cell value that is, `dp[i][j]=dp[i-1][j]`.
+- Case2- If we consider the ith coin, then the cell value is `1+ dp[i][j-coins[i-1]]` cell value.
+- The final value of the cell is the minimum of both the cases,that is, `dp[i][j]= min(dp[i-1][j], 1+dp[i][j-coins[i-1]])`.
+- The final answer will be `dp[n][amount]`.
+
+## Code
+
+import Tabs from "@theme/Tabs";
+import TabItem from "@theme/TabItem";
+
+<Tabs
+  defaultValue="java"
+  values={[
+    { label: 'Java', value: 'java', },
+    { label: 'Python', value: 'py', },
+  ]
+}>
+
+<TabItem value="java">
+
+```java
+import java.util.Scanner;
+
+class CoinChangeUsingDp {
+
+    public static int coinChange(int[] coins, int amount) {
```

Comment:
Can you please comment this section of your code till your main function.

Reply:
![image](https://user-images.githubusercontent.com/64724041/119080890-88adb900-ba18-11eb-9457-14a35d08b141.png)


Please could you tell me whether this is correct?  Or please could you specify the line numbers which I have to comment out.
Thank you

Reply:
I have added the comments now.